### PR TITLE
Process results

### PR DESCRIPTION
**Pull request by amandarichardsonn**
1. Process Results implemented for all scaling tests EXCPET colocated inference (avoiding conflicts with Andrew PR merge)
2. Recommendations for error checking would be cool
3. PROCESS RESULTS WILL NOT WORK UNLESS ALL RUN FOLDERS R VALID
4. 2 graphs produced for cpp and fortran (fortran scaling test in andrew PR)
5. iteration flag implemented for inference tests
6. wall-time flag implemented for all standard tests
7. db cpu list support for throughput scaling

### CODE REVIEW

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 1]
```diff
@@ -146,7 +153,7 @@ void run_mnist(const std::string& model_name,
     std::cout<<"All ranks have Resnet image"<<std::endl;
 
   double loop_start = MPI_Wtime();
-  for (int i = 0; i < 101; i++) {
+  for (int i = 0; i < iterations; i++) {
```
Comment (billschereriii) [inline cpp-inference/inference_scaling_imagenet.cpp:156]: There is one extra iteration done to warm up caches, so this should actually be `iterations + 1`

File: cpp-throughput/throughput_scaling.cpp [hunk 2]
```diff
@@ -61,7 +61,7 @@ void run_throughput(std::ofstream& timing_file,
         client.put_tensor(key,
                           array.data(),
                           {1, n_values},
-                          SRTensorTypeFloat,
+                          SRTensorTypeFloat,
```
Comment (billschereriii) [inline cpp-throughput/throughput_scaling.cpp:64]: Please try not to leave stray whitespace at the end of lines

File: driverdataaggregation/main.py [hunk 3]
```diff
@@ -1,5 +1,8 @@
-import fire
 from utils import *
+from driverprocessresults.main import *
+
+if __name__ == "__main__":
```
Comment (billschereriii) [inline driverdataaggregation/main.py:4]: Please add a comment explaining why this was done

File: driverdataaggregation/main.py [hunk 4]
```diff
@@ -14,19 +17,19 @@ def aggregation_scaling(self,
...
-                            tensor_bytes=[1024, 8192, 16384, 32769, 65538,
-                                          131076, 262152, 524304, 1024000,
-                                          2048000],
-                            tensors_per_dataset=[1,2,4],
-                            client_threads=[1,2,4,8,16,32],
-                            cpu_hyperthreads=2):
+                            clients_per_node=[48],
+                            client_nodes=[48],
+                            iterations=3,
+                            tensor_bytes=[1024],
```
Comment (billschereriii) [inline driverdataaggregation/main.py:27]: These don't look like good defaults -- were they leftovers from a scaled-back testing run?

File: driverdataaggregation/main.py [hunk 5]
```diff
@@ -93,11 +112,12 @@ def aggregation_scaling(self,
...
                                 run_db_as_batch,
-                                db_hosts)
+                                db_hosts,
+                                wall_time)
 
 
-            for c_nodes, cpn, _bytes, t_per_dataset, c_threads in product(
-                client_nodes, clients_per_node, tensor_bytes, tensors_per_dataset, client_threads
+            for c_nodes, cpn, _bytes, t_per_dataset, c_threads, language in product(
+                client_nodes, clients_per_node, tensor_bytes, tensors_per_dataset, client_threads, languages
```
Comment (billschereriii) [inline driverdataaggregation/main.py:120]: It would probably be good to convert the language to lower-case here so that Fortran, FORTRAN, fortran are all accepted as well as cpp, CPP, Cpp.

File: driverdataaggregation/main.py [hunk 6]
```diff
@@ -327,19 +352,19 @@ def aggregation_scaling_python(self,
...
-                                client_threads=[1,2,4,8,16,32],
-                                cpu_hyperthreads=2):
+                                client_nodes=[128],
+                                iterations=1,
+                                tensor_bytes=[1024],
+                                tensors_per_dataset=[4],
+                                client_threads=[32],
+                                cpu_hyperthreads=2,
+                                languages=["cpp"],
+                                wall_time="05:00:00"):
```
Comment (billschereriii) [inline driverdataaggregation/main.py:367]: Here too, please set good defaults

File: driverdataaggregation/main.py [hunk 7]
```diff
@@ -386,12 +411,29 @@ def aggregation_scaling_python(self,
                                     to request that the consumer application utilizes
                                     all physical cores for each client thread.
             :type cpu_hyperthreads: int, optional
+            :param languages: which language to use for the tester "cpp" or "fortran"
```
Comment (billschereriii) [inline driverdataaggregation/main.py:414]: Wording implies it's one or the other, but your code supports running with both

File: driverdataaggregation/main.py [hunk 8]
```diff
@@ -74,12 +77,28 @@ def aggregation_scaling(self,
                                  to request that the consumer application utilizes
                                  all physical cores for each client thread.
         :type cpu_hyperthreads: int, optional
+        :param languages: which language to use for the tester "cpp" or "fortran"
```
Comment (billschereriii) [inline driverdataaggregation/main.py:80]: Wording implies it's one or the other, but your code supports running with both

File: driverdataaggregation/main.py [hunk 9]
```diff
@@ -386,12 +411,29 @@ def aggregation_scaling_python(self,
...
             """
             logger.info("Starting dataset aggregation scaling with python tests")
             logger.info(f"Running with database backend: {get_db_backend()}")
             logger.info(f"Running with launcher: {launcher}")
 
-            exp = create_folder(exp_name, launcher)
+            exp, result_path = create_folder(exp_name, launcher)
+            
+            write_run_config(result_path,
+                    colocated=0,
```
Comment (billschereriii) [inline driverdataaggregation/main.py:426]: Is this intended for future use?

File: driverdataaggregation/main.py [hunk 10]
```diff
@@ -487,14 +532,13 @@ def aggregation_scaling_python_fs(self,
...
-                                          131076, 262152, 524304, 1024000,
-                                          2048000],
-                            tensors_per_dataset=[1,2,4],
-                            client_threads=[1,2,4,8,16,32],
-                            cpu_hyperthreads=2):
+                            tensor_bytes=[1024],
+                            tensors_per_dataset=[4],
+                            client_threads=[16,32],
+                            cpu_hyperthreads=2,
+                            languages=["cpp"]):
```
Comment (billschereriii) [inline driverdataaggregation/main.py:541]: Need good defaults

File: driverdataaggregation/main.py (hunk 5 above)
Comment (billschereriii) [reply driverdataaggregation/main.py:120]: (apply this to the other tests as well)

File: driverdataaggregation/main.py [hunk 11]
```diff
@@ -526,29 +570,41 @@ def aggregation_scaling_python_fs(self,
                                  to request that the consumer application utilizes
                                  all physical cores for each client thread.
         :type cpu_hyperthreads: int, optional
+        :param languages: which language to use for the tester "cpp" or "fortran"
```
Comment (billschereriii) [inline driverdataaggregation/main.py:573]: Update wording

File: driverinference/main.py [hunk 12]
```diff
@@ -65,19 +67,42 @@ def inference_standard(self,
         :type client_nodes: list[int], optional
         :param rebuild_model: force rebuild of PyTorch model even if it is available
         :type rebuild_model: bool
+        :param iterations: number of put/get loops run by the applications
+        :type iterations: int
+        :param languages: which language to use for the tester "cpp" or "fortran"
```
Comment (billschereriii) [inline driverinference/main.py:72]: Update description

File: driverinference/main.py [hunk 13]
```diff
@@ -117,12 +145,13 @@ def inference_standard(self,
...
             if stat[0] != status.STATUS_COMPLETED:
                 logger.error(f"One of the scaling tests failed {infer_session.name}")
+        self.process_scaling_results(scaling_results_dir=exp_name)
   
     def inference_colocated(self,
                             exp_name="inference-colocated-scaling",
                             node_feature={"constraint": "P100"},
                             launcher="auto",
-                            nodes=[12],
+                            nodes=[1],
```
Comment (billschereriii) [inline driverinference/main.py:154]: not a good default

File: driverinference/main.py [hunk 14]
```diff
@@ -131,7 +160,7 @@ def inference_colocated(self,
                             batch_size=[1],
                             device="GPU",
                             num_devices=1,
-                            net_ifname="lo",
+                            net_ifname="ipogif0",
```
Comment (billschereriii) [inline driverinference/main.py:163]: I think this name is probably very specific to Horizon

File: driverprocessresults/main.py [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+            for run_folder in run_folders:
+                
+                if 'run' in run_folder:
+                    session_folders += ["results/" + scaling_results_dir + "/" + run_folder + "/" + d for d in os.listdir("results/" + scaling_results_dir + "/" + run_folder) if "sess" in d]
+                    run_list += [Path("results/" + scaling_results_dir + "/" + run_folder)]
+            try:
+                # write csv each so this function is idempotent
+                # csv's will not be written if they are already created
+                #tqdm creates a smart progress bar for the loops
+                for session_folder in tqdm(session_folders, desc="Processing scaling results...", ncols=80): #QUESTION: ncols: hardcoded?
```
Comment (billschereriii) [inline driverprocessresults/main.py:45]: terminals tend to be 80 characters wide. Probably shouldn't have this comment in production code

File: driverprocessresults/main.py [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+                    # want to catch all exceptions and skip runs that may
+                    # not have completed or finished b/c some reason i.e. node failure
+                    except Exception as e:
+                        logger.warning(f"Skipping {session_folder} could not process results")
+                        logger.error(e)
+                        continue
+                # collect all written csv into dataframes to concat
+                for run in tqdm(run_list, desc="Creating plots...", ncols=80):
+                    self._other_plots(run, scaling_results_dir) #need to change this to an fn
+                for session in tqdm(session_folders, desc="Collecting scaling results...", ncols=80): #QUESTION: ncols: hardcoded?
```
Comment (billschereriii) [inline driverprocessresults/main.py:57]: same as before

File: driverprocessresults/main.py [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+                        session_name = os.path.basename(session)
+                        split = os.path.dirname(session)
+                        run_name = os.path.basename(split)
+                        stats_path = os.path.join(final_stat_dir, run_name, session_name, session_name + ".csv")
+                        run_df = pd.read_csv(str(stats_path))
+                        dataframes.append(run_df)
+                    
+                    # catch all and skip for reason listed above
+                    except Exception as e:
+                        logger.warning(f"Skipping path {session} could not read results csv")
```
Comment (billschereriii) [inline driverprocessresults/main.py:68]: maybe put a colon (`:`) after {session}?

File: driverprocessresults/main.py [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+        run_name = os.path.basename(split)
+        all_stats_dir = cls._create_stats_dir(session_path)
+        session_stats_dir = all_stats_dir / run_name / session_name
+        
+        if delete_previous and session_stats_dir.is_dir():
+            shutil.rmtree(session_stats_dir)
+        
+        if not session_stats_dir.is_dir():
+            os.makedirs(session_stats_dir)
+            #HERE
```
Comment (billschereriii) [inline driverprocessresults/main.py:93]: Here?

File: driverprocessresults/main.py [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+            for file in files: 
+                if '.csv' in file:
+                    #creating fp=sess/file.csv
+                    fp = os.path.join(session_path, file)
+                    #opening csv file
+                    with open(fp) as f:
+                        #scan over all lines in file
+                        for i, line in enumerate(f):
+                            #QUESTION: ask for an explanation
+                            #does this created 2 keys and assigns all values to the 2 keys?
```
Comment (billschereriii) [inline driverprocessresults/main.py:105]: A few more comments that can probably be deleted in these last few lines

File: driverprocessresults/main.py [hunk 20]
```diff
@@ -0,0 +1,208 @@
...
+                            vals = line.split(',')
+                            if vals[1] in function_times.keys():
+                                function_times[vals[1]].append(float(vals[2]))
+                            else:
+                                function_times[vals[1]] = [float(vals[2])]
+                else:
+                    if verbose: #QUESTION: what is the meaning? right now verbose will always be false?
+                        print(file) #why do we print file()?
```
Comment (billschereriii) [inline driverprocessresults/main.py:113]: These sound like good questions to resolve before merging the PR :)

File: driverprocessresults/main.py [hunk 21]
```diff
@@ -0,0 +1,208 @@
...
+                if "client()" in function_times:
+                    cls._make_hist_plot(function_times['client()'], 'client()', 'client_constructor_dist.pdf', session_stats_dir)
+                
+                if "put_tensor" in function_times:
+                    cls._make_hist_plot(function_times['put_tensor'], 'put_tensor()', 'put_tensor.pdf', session_stats_dir)
+
+                if "unpack_tensor" in function_times:
+                    cls._make_hist_plot(function_times['unpack_tensor'], 'unpack_tensor()', 'unpack_tensor.pdf', session_stats_dir)
+
+                if "get_list" in function_times:
```
Comment (billschereriii) [inline driverprocessresults/main.py:135]: Seems like there should be a better way to do these than hard-coding each potential name. Failing that, making a list of the names and looping through it would eliminate a lot of duplicate code here

File: driverthroughput/main.py [hunk 22]
```diff
@@ -1,26 +1,30 @@
-import fire
 from utils import *
+from driverprocessresults.main import *
+
+if __name__ == "__main__":
+    sys.path.append("..")
```
Comment (billschereriii) [inline driverthroughput/main.py:5]: Again, please comment what this is here for

File: driverthroughput/main.py [hunk 23]
```diff
@@ -1,26 +1,30 @@
...
 
 from smartsim.log import get_logger, log_to_file
 logger = get_logger("Scaling Tests")
 
 class Throughput:
     def throughput_standard(self,
-                           exp_name="throughput-standard-scaling",
+                           exp_name="inference-standard-scaling",
```
Comment (billschereriii) [inline driverthroughput/main.py:12]: this is throughput, not inference. Why the change?

File: driverthroughput/main.py [hunk 24]
```diff
@@ -1,26 +1,30 @@
...
                            clients_per_node=[32],
-                           client_nodes=[128, 256, 512],
-                           iterations=100,
-                           tensor_bytes=[1024, 8192, 16384, 32768, 65536, 131072,
-                                         262144, 524288, 1024000, 2048000, 4096000]):
+                           client_nodes=[128],
+                           iterations=3,
+                           tensor_bytes=[1024],
+                           languages=["cpp"],
+                           wall_time="05:00:00"):
```
Comment (billschereriii) [inline driverthroughput/main.py:27]: Please set good defaults

File: driverthroughput/main.py [hunk 25]
```diff
@@ -53,41 +57,57 @@ def throughput_standard(self,
         :type iterations: int
         :param tensor_bytes: list of tensor sizes in bytes
         :type tensor_bytes: list[int], optional
+        :param languages: which language to use for the tester "cpp" or "fortran"
```
Comment (billschereriii) [inline driverthroughput/main.py:60]: text implies single language only

File: driverthroughput/main.py [hunk 26]
```diff
@@ -53,41 +57,57 @@ def throughput_standard(self,
...
+                    client_per_node=clients_per_node,
+                    client_nodes=client_nodes,
+                    database_nodes=db_nodes,
+                    database_cpus=db_cpus,
+                    iterations=iterations,
+                    tensor_bytes=tensor_bytes,
+                    language=languages,
+                    wall_time=wall_time)
+        permss = list(product(db_nodes, db_cpus))
+        for permsss in permss:
```
Comment (billschereriii) [inline driverthroughput/main.py:81]: can you pick a better variable name than `permsss` please?

File: driverthroughput/main.py [hunk 27]
```diff
@@ -53,41 +57,57 @@ def throughput_standard(self,
...
             # start the database only once per value in db_nodes so all permutations
             # are executed with the same database size without bringin down the database
             db = start_database(exp,
                                 db_node_feature,
                                 db_port,
-                                db_node_count,
-                                db_cpus,
+                                dbn,
+                                dbc,
                                 None, # not setting threads per queue in throughput tests
```
Comment (billschereriii) [inline driverthroughput/main.py:90]: Switching to less descriptive variable names will make it harder for the next person to maintain this code

File: driverthroughput/main.py [hunk 28]
```diff
@@ -162,22 +185,23 @@ def _create_throughput_session(cls,
...
                            net_ifname="lo",
-                           clients_per_node=[3],
+                           clients_per_node=[48],
                            pin_app_cpus=[False],
-                           iterations=100,
-                           tensor_bytes=[1024, 8192, 16384, 32768, 65536, 131072,
-                                         262144, 524288, 1024000, 2048000, 4096000]):
+                           iterations=3,
+                           tensor_bytes=[1024],
+                           languages=["cpp"]):
```
Comment (billschereriii) [inline driverthroughput/main.py:204]: Double-check these are good defaults. They look better than some of the others

File: driverthroughput/main.py [hunk 29]
```diff
@@ -206,16 +230,27 @@ def throughput_colocated(self,
         :type iterations: int
         :param tensor_bytes: list of tensor sizes in bytes
         :type tensor_bytes: list[int], optional
+        :param languages: which language to use for the tester "cpp" or "fortran"
```
Comment (billschereriii) [inline driverthroughput/main.py:233]: which language`s`?

File: driverthroughput/main.py [hunk 30]
```diff
@@ -286,6 +324,7 @@ def _create_colocated_throughput_session(cls,
 
         name = "-".join((
             "throughput-sess-colo",
+            str(language),
```
Comment (billschereriii) [inline driverthroughput/main.py:327]: if language is already "cpp" or "fortran", you don't need to convert it to a string, right?

File: driverthroughput/main.py [hunk 31]
```diff
@@ -315,5 +354,10 @@ def _create_colocated_throughput_session(cls,
...
                     iterations=iterations,
-                    tensor_bytes=_bytes)
-        return model
\ No newline at end of file
+                    tensor_bytes=_bytes,
+                    language=language)
+        return model
+
+if __name__ == "__main__":
+    import fire
```
Comment (billschereriii) [inline driverthroughput/main.py:362]: this is probably purely stylistic, but I like having all the imports together at the top of the file

File: utils.py [hunk 32]
```diff
@@ -189,7 +190,8 @@ def write_run_config(path, **kwargs):
         "smartsim_version": smartsim.__version__,
         "smartredis_version": "0.3.1", # TODO put in smartredis __version__
         "db": get_db_backend(),
-        "date": str(get_date())
+        "date": str(get_date()),
+        "language": "cpp"
```
Comment (billschereriii) [inline utils.py:194]: no fortran support here?

File: utils.py [hunk 33]
```diff
@@ -213,4 +215,4 @@ def get_db_backend():
     :rtype: str
     """
     db_backend_path = smartsim._core.config.CONFIG.database_exe
-    return os.path.basename(db_backend_path)
+    return os.path.basename(db_backend_path) #Will need to look into this further, will need to return a list
```
Comment (billschereriii) [inline utils.py:218]: Looks like a todo item that got missed

**Review by billschereriii — CHANGES_REQUESTED** [review]
Thanks for your hard work on this PR.
This is generally looking really solid. There's a bit of debugging commentary and settings that need to be adjusted, but I'm not seeing any huge problems that will take a long time to sort out.

File: driverdataaggregation/main.py (hunk 7 above)
Comment (amandarichardsonn) [reply driverdataaggregation/main.py:414]: changed to and/or

File: driverdataaggregation/main.py (hunk 8 above)
Comment (amandarichardsonn) [reply driverdataaggregation/main.py:80]: changed to and/or

File: driverdataaggregation/main.py (hunk 9 above)
Comment (amandarichardsonn) [reply driverdataaggregation/main.py:426]: The write run config at the beginning is placed in the top level directory of the run folder -> its so that users may go in and look at what flags they set for that run instead of going through each of the session folders and adding up all permutations

File: driverdataaggregation/main.py (hunk 11 above)
Comment (amandarichardsonn) [reply driverdataaggregation/main.py:573]: changed to and/or

File: driverinference/main.py (hunk 12 above)
Comment (amandarichardsonn) [reply driverinference/main.py:72]: changed to and/or

File: driverinference/main.py (hunk 14 above)
Comment (amandarichardsonn) [reply driverinference/main.py:163]: Hmmm Ill consult Andrew for what to put there or look back at the original name

File: driverprocessresults/main.py (hunk 17 above)
Comment (amandarichardsonn) [reply driverprocessresults/main.py:68]: I agree

File: driverthroughput/main.py (hunk 23 above)
Comment (amandarichardsonn) [reply driverthroughput/main.py:12]: thank you for catching that!

File: driverthroughput/main.py (hunk 26 above)
Comment (amandarichardsonn) [reply driverthroughput/main.py:81]: yes MERP nice catch

File: driverthroughput/main.py (hunk 27 above)
Comment (amandarichardsonn) [reply driverthroughput/main.py:90]: I agree, I will change

File: driverprocessresults/main.py [hunk 34]
```diff
@@ -0,0 +1,208 @@
...
+            dataframes = []
+            final_stat_dir = "results" / Path(scaling_results_dir) / "stats"
+            run_folders = os.listdir("results" / Path(scaling_results_dir))
+            session_folders = []
+            run_list = []
+            for run_folder in run_folders:
+                
+                if 'run' in run_folder:
+                    session_folders += ["results/" + scaling_results_dir + "/" + run_folder + "/" + d for d in os.listdir("results/" + scaling_results_dir + "/" + run_folder) if "sess" in d]
+                    run_list += [Path("results/" + scaling_results_dir + "/" + run_folder)]
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply driverprocessresults/main.py:41]: done

File: driverprocessresults/main.py (hunk 21 above)
Comment (amandarichardsonn) [reply driverprocessresults/main.py:137]: done!

File: driverthroughput/main.py (hunk 31 above)
Comment (amandarichardsonn) [reply driverthroughput/main.py:362]: done

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 35]
```diff
@@ -86,30 +93,41 @@ void run_mnist(const std::string& model_name,
...
-          std::string script_key = "resnet_script_" + std::to_string(i);
           client.set_model_from_file(model_key, model_filename, "TORCH", device, batch_size);
           client.set_script_from_file(script_key, device, "./data_processing_script.txt");
+      }
+    }
+    if(is_colo && rank % n_clients == 0) {
+      std::cout<<"Setting Resnet Model from scaling app" << std::endl << std::flush;
+      std::cout<<"Setting with batch_size: " << std::to_string(batch_size) << std::endl << std::flush;
+      std::cout<<"Setting on device: " << device << std::endl << std::flush;
+      std::cout<<"Setting on " << std::to_string(num_devices) << " devices" <<std::endl << std::flush;
```
Comment (billschereriii) [inline cpp-inference/inference_scaling_imagenet.cpp:121]: one std::flush at the end of the last print statement is sufficient. There's no need to flush after each one

File: driverthroughput/main.py [hunk 36]
```diff
@@ -2,29 +2,40 @@
 from driverprocessresults.main import *
 
 if __name__ == "__main__":
+    """ Takes the pwd, then navigates to the root to append packages.
+    Python is then able to find our *.py files in that directory.
+    """
     sys.path.append("..")
 
+if __name__ == "__main__":
```
Comment (billschereriii) [inline driverthroughput/main.py:10]: Here, too, you have duplicate `if __name__ == "__main__":`

Comment (amandarichardsonn) [reply driverthroughput/main.py:10]: answered above!

File: driverinference/main.py [hunk 37]
```diff
@@ -1,9 +1,20 @@
...
 from driverprocessresults.main import *
+import time
 
 if __name__ == "__main__":
+    """ Takes the pwd, then navigates to the root to append packages.
+    Python is then able to find our *.py files in that directory.
+    """
     sys.path.append("..")
 
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply driverinference/main.py:11]: okay so I asked matt drozt, he said that we will need sys.append() to run before the import utils, so I actually made a mistake with having it after and thank u for catching that!! He says "if you are executing directly via driver*/main.py instead of the driver.py script thus the multiple  if __name__ == '__main__': blocks. you could let sys.path.append(...) run unconditionally, but that will pollute the PYTHONPATH when executing driver.py directly. That said, it wont break anything so its not that big a deal"

should I combine the two still?

File: cpp-inference/inference_scaling_imagenet.cpp (hunk 35 above)
Comment (amandarichardsonn) [reply cpp-inference/inference_scaling_imagenet.cpp:121]: done!

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 38]
```diff
@@ -86,30 +93,41 @@ void run_mnist(const std::string& model_name,
   int num_devices = get_num_devices();
   bool use_multigpu = (0 == device.compare("GPU")) && num_devices > 1;
   bool should_set = get_set_flag();
+  std::string model_key = "resnet_model";
+  std::string script_key = "resnet_script";
   if (should_set) {
     int batch_size = get_batch_size();
     int n_clients = get_client_count();
-    if (rank == 0) {
+    if (is_colo == false && rank == 0) {
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply cpp-inference/inference_scaling_imagenet.cpp:101]: okay facts I see where this coulda gone wrong

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 39]
```diff
@@ -38,6 +38,12 @@ bool get_set_flag() {
   return should_set;
 }
 
+bool get_colo() {
+  char* is_colocated = std::getenv("SS_COLO");
+  bool is_colo = is_colocated ? std::stoi(is_colocated) : false;
+  return is_colo;
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply cpp-inference/inference_scaling_imagenet.cpp:44]: okay yes that is true, if im passing in a bool i dont need to convert

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 40]
```diff
@@ -38,6 +38,12 @@ bool get_set_flag() {
   return should_set;
 }
 
+bool get_colo() {
+  char* is_colocated = std::getenv("SS_COLO");
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply cpp-inference/inference_scaling_imagenet.cpp:42]: It was the SS_COLO that I was referring to

File: driverinference/main.py (hunk 37 above)
[the remark this answers is not in the document]
Comment (billschereriii) [reply driverinference/main.py:11]: I see. Sounds good then

**Review by billschereriii — APPROVED** [review]
LGTM! Thanks for all your hard work getting us to this point!

**Comment by amandarichardsonn** [comment]
db_cpu list support, wall_time flag, iteration flag, automated graph creation, invalid folder creation ( #22 #24 #23 #26 #21 )

This ticket adds db_cpu list support for throughput scaling, an iteration for inference scaling, wall time flag for all std scaling tests and all graph creation has been move to a single plotter folder. An additional plot flag has been added to allow users to decide what flag to plot against. The invalid folder creation bug was fixed during another fix. Additional edits have been made to the cpp file to pass inference tests.
